source/opt/inst_bindless_check_pass.h

```diff
@@ -84,24 +89,88 @@ class InstBindlessCheckPass : public InstrumentPass {
+  // array |var_id| from debug input buffer and return id of value.
+  uint32_t GenDebugReadLength(uint32_t var_id, InstructionBuilder* builder);
+
+  // Generate instructions into |builder| to read initialization status of
+  // descriptor array |image_id| at |index_id| from debug input buffer and
+  // return id of value.
+  uint32_t GenDebugReadInit(uint32_t image_id, uint32_t index_id,
+                            InstructionBuilder* builder);
+
+  // Clone original reference
+  uint32_t CloneOriginalReference(InstructionBuilder* builder);
+
+  // If |inst| references through a descriptor, return the id of the value it
+  // references through. Else return 0.
+  uint32_t GetDescriptorValueId(Instruction* inst);
+
+  // Analyze descriptor reference |ref_inst| and save components for later use.
+  bool AnalyzeDescriptorReference(Instruction* ref_inst);
+
+  // Generate instrumentation code for generic test result |check_id|, starting
+  // with |builder| of block |new_blk_ptr|, adding new blocks to |new_blocks|.
+  // Use |error_id|, length_id and |stage_idx| (among other saved info) to
+  // create debug output. Generate conditional branch to a valid or invalid
+  // branch. Generate valid block which does original reference. Generate
+  // invalid block which writes debug output. Generate merge block for valid
+  // and invalid branches. Kill original reference.
+  void GenCheckCode(uint32_t check_id, uint32_t error_id, uint32_t length_id,
+                    uint32_t stage_idx, InstructionBuilder* builder,
+                    std::unique_ptr<BasicBlock>* new_blk_ptr,
+                    std::vector<std::unique_ptr<BasicBlock>>* new_blocks);
+
+  // Initialize state for instrumenting bindless checking
+  void InitializeInstBindlessCheck();
+
+  // Apply GenBoundsCheckCode to every instruction in module. Then apply
+  // GenInitCheckCode to every instruction in module.
   Pass::Status ProcessImpl();
 
   // True if VK_EXT_descriptor_indexing is defined
   bool ext_descriptor_indexing_defined_;
 
-  // Enable instrumentation of runtime arrays
-  bool runtime_array_enabled_;
+  // Enable instrumentation of runtime array length checking
+  bool input_length_enabled_;
+
+  // Enable instrumentation of descriptor initialization checking
+  bool input_init_enabled_;
 
   // Mapping from variable to descriptor set
   std::unordered_map<uint32_t, uint32_t> var2desc_set_;
 
   // Mapping from variable to binding
   std::unordered_map<uint32_t, uint32_t> var2binding_;
+
+  // Descriptor reference analysis components, both ids and instructions
+  //
+  // Image value
+  uint32_t image_id_;
+  Instruction* image_inst_;
+  // Descriptor load
+  uint32_t load_id_;
+  Instruction* load_inst_;
+  // Descriptor pointer
+  uint32_t ptr_id_;
+  Instruction* ptr_inst_;
+  // Base descriptor variable
+  uint32_t var_id_;
+  Instruction* var_inst_;
+  // Index into descriptor array
+  uint32_t index_id_;
+  // Original reference
+  Instruction* ref_inst_;
```

Comment:
I'm wondering if you really need all of these values.  When I look at `AnalyzeDescriptorReference`, it seems like many of these are readily available from others.  For example, `load_id_` is the result id of `load_inst_`.  Is there a good reason to keep both?

Also `ptr_id_` is used in just a single function.  Does it need to be a member variable?

`ptr_inst_` is used in just in one other place?

It might be worth writing a separate member function that retrieves these values by looking at `load_inst_` instead of storing all of them.  I don't think it will make much difference in the compile time, and it because easier to make sense of the state of the class being carried around.

Reply:
Now that I have looked at more code, I am wondering if these should just be their own struct that is returned by `AnalyzeDescriptorReference`.  It seems like `AnalyzeDescriptorReference` is called, and then the member are essentially dead at the end of the function that calls it.  This just gets confusing make it part of the state of the the pass.  I would have expected them to be used in more places.

Reply:
OK. I have cleaned up the "redundant" members and have encapsulated the remaining members into a struct whose logical and storage lifetimes will match.